### DISCUSSION TITLE
How to build multi-tenancy?

### DISCUSSION THREAD

**Discussion by rohanrajpal** [vendurehq/vendure#495] [Q&A] [answered]
Hey, I was reading about multi-tenancy here
https://www.vendure.io/docs/developer-guide/channels/
There was a mention of some community projects trying to do multi-tenancy, but I couldn't find any. Can someone please share such a project?

Also, I read [here](https://pagepro.co/blog/multi-tenant-ecommerce-platform-in-three-months/) in different ways to add multi-tenancy support. 

I want to go for the "one schema per tenant" solution but using Vendure as my backend. Does anyone have any library in mind similar to [django-tenants](https://github.com/django-tenants/django-tenants) to add multi-vendor support?

Thanks

**Comment by michaelbromley** [comment, marked as answer]
Hi there,

I skimmed the article you linked but I don't really understand what is meant by "one schema per tenant", but it seems it may be a Postgres-specific solution? Vendure supports several different DBs so I doubt any DB-level solution would work here.

By "community projects" it is referring to contributions to the core that enable multi-tenancy features, rather than a plugin you can install. For example:

* [feat(core): Make Customers ChannelAware](https://github.com/vendure-ecommerce/vendure/pull/454)
* [Making Orders ChannelAware](https://github.com/vendure-ecommerce/vendure/pull/454)

These PRs make it so Orders and Customers are assigned to specific channels. So as of now the following entities are "channel-aware":

* Order
* Customer
* Role
* Product
* Collection
* Shipping Method

A Channel in this case can be used to represent a tenant (vendor). So currently it _may_ be possible to have a multi-tenant solution, but I'm pretty sure you'll find some rough edges and cases where things may break down or not be as expected. 

### Alternative solution

If you are aiming to build a Shopify-like platform where a merchant can create an account and then get a new e-commerce store created for him or her, then there is another approach you may want to consider.

You can do it at the operations level by creating infrastructure that will create a new Vendure instance for each tenant. In this way each instance is truly isolated. 

Pros:
* Works right now
* Less complexity in the application

Cons:
* More complexity in the infrastructure

**Reply by rohanrajpal** [reply]
Thanks for the quick reply! Yeah, I was planning to build a Shopify like platform and found the isolated approach to be infra heavy. I think I'll go with the channels method for now and will make some PRs if I solve any issues!

**Reply by martijnvdbrug** [reply]
I am also using Vendure as multi-tenant marketplace: https://github.com/martijnvdbrug/shops/tree/master/vendure 
There are some minor issues like assets and facets being shared, but no serious issues so far. Take a look at the repo for an example on multi-tenant payments

**Reply by rohanrajpal** [reply]
Thanks for sharing @martijnvdbrug! Having a look right away.

**Reply by mariusa** [reply]
> aiming to build a Shopify-like platform where a merchant can create an account and then get a new e-commerce store created for him or her,

What would be the work required to support this built-in? Listing some tasks would make it easier for potential contributors to tackle some of them.

**Comment by rohanrajpal** [comment]
@michaelbromley 
So the current requirement for what I'm building will be more of a pseudo-multi-tenant solution.
The flow of my product will be like
- user downloads our app
- they make an account using their mobile number and sign in with an OTP
- they choose a Name for the store
- a store is created for them, but it's just a new page on an already deployed website
  - so if the store name is 'John Doe' and lets stay our app name is 'StoreMaker'
  - the website for the new user will be `storemaker.com/john-doe`
  - later if another user registers a store with the name 'Rohan Rajpal', his store link will be `storemaker.com/rohan-rajpal`
  - if someone creates a store with the same name, we'll modify the link like `storemaker.com/rohan-rajpal1`

To implement this I'm thinking of creating a new channel every time a new user comes in and creates a new channel page. I hope this approach would be fine?

**Reply by michaelbromley** [reply]
Hi, it sounds reasonable. Some things to consider:

* Will the merchants be interacting with Vendure solely through your app? I.e. they do not need to use the web-based Admin UI? If so, then it should be fine
* Will merchants expect to be able to set prices in multiple currencies? If so then this might get a bit tricky.

**Reply by rohanrajpal** [reply]
Hey, thanks for the quick reply

> Hi, it sounds reasonable. Some things to consider:
> 
> * Will the merchants be interacting with Vendure solely through your app? I.e. they do not need to use the web-based Admin UI? If so, then it should be fine

Yes, they will be interacting only through the app.

> * Will merchants expect to be able to set prices in multiple currencies? If so then this might get a bit tricky.

No, there is only a single currency for now.

I think I should start working on the app then

**Reply by rohanrajpal** [reply]
Hey @michaelbromley, it would be great if you could help me out with the following 

I want to build more of a website that offers booking services, specifically for salons, but is similar to an e-commerce in terms of orders, carts, payments, etc.
#### Frontend
- I plan to use the Next.js commerce storefront and make it work with Vendure
  - Is this a good approach? I did see in slack that Vendure is also trying to add support for nextjs commerce. So thought this might be right thing to start with
- Every new store will be a new page
  - For example, `storemaker.com/awesome-sneakers` is a store and will have its code in `src/pages/awesome-sneakers.tsx`
#### Backend
- When the user with the name `John Doe` signs up, a new channel is created, let's say `john-doe-channel`
- Subsequently, a new role is created for that channel. Lets call it `john-doe-role`
- Then, the user enters a name for the store, which is `Awesome Sneakers`
  -  This triggers a script that builds a new page which would use `john-doe-channel` to display products
  - A little confusion i have is how to fetch products specific to a channel? There are options in the docs to update the current channel, but how do i make these pages channel specific?
- The user would only login through their mobile number and no password, they would get an OTP everytime they have to login, email's are not that widely used in my country (India)
- To add/modify inventory, orders, etc. of the `john-doe-channel`, we will use the `john-doe-role`

If you have any feedback or find something incorrect with the above approach, please let me know!

**Reply by michaelbromley** [reply]
Hi, 

>I plan to use the Next.js commerce storefront and make it work with Vendure

Next.js is a good choice - many people are already using it with Vendure. The Next.js Commerce project still seems a bit immature to recommend - it is still tightly coupled to BigCommerce for example.

>Every new store will be a new page

I don't quite understand the need for this part. Why can you not have a single page that is used by all stores, where the data is dynamically populated from the API? A new physical file for each store seems highly unusual.

> A little confusion i have is how to fetch products specific to a channel? There are options in the docs to update the current channel, but how do i make these pages channel specific?

The Admin API has the [assignProductsToChannel](https://www.vendure.io/docs/graphql-api/admin/mutations/#assignproductstochannel) mutation which you can use to assign a product to the channel of a particular merchant. Each Channel has a `token` which is a unique string used to identify the channel. If you pass it to the API as a `vendure-token` header as described [here](https://www.vendure.io/docs/storefront/shop-api-guide/#universal-parameters), you will get content specific to that channel.

The other parts sound fine. I'll be interested to see the finished product!

**Reply by rohanrajpal** [reply]
> Hi,
> 
> > I plan to use the Next.js commerce storefront and make it work with Vendure
> 
> Next.js is a good choice - many people are already using it with Vendure. The Next.js Commerce project still seems a bit immature to recommend - it is still tightly coupled to BigCommerce for example.

Yeah, I found it tough to set up too. Any other Nextjs based framework or storefront that you would recommend? 

Starting from scratch would be a lot of work otherwise. I was thinking of using [react-storefront](https://github.com/storefront-foundation/react-storefront). Going by their documentation, I only need to write a connector to integrate Vendure.
> 
> > Every new store will be a new page
> 
> I don't quite understand the need for this part. Why can you not have a single page that is used by all stores, where the data is dynamically populated from the API? A new physical file for each store seems highly unusual.
> 

So I want the creators of the stores to easily share the link of their store. Their customers can open the store link with any browser and place their orders. If the link is the same, how would we figure out which store to display to the customer?

Some examples of products following the same page method
- https://play.google.com/store/apps/details?id=com.khatabook.dukaan&hl=en_IN&gl=US
- https://play.google.com/store/apps/details?id=com.bikayi.android&hl=en_IN&gl=US

> > A little confusion i have is how to fetch products specific to a channel? There are options in the docs to update the current channel, but how do i make these pages channel specific?
> 
> The Admin API has the [assignProductsToChannel](https://www.vendure.io/docs/graphql-api/admin/mutations/#assignproductstochannel) mutation which you can use to assign a product to the channel of a particular merchant. Each Channel has a `token` which is a unique string used to identify the channel. If you pass it to the API as a `vendure-token` header as described [here](https://www.vendure.io/docs/storefront/shop-api-guide/#universal-parameters), you will get content specific to that channel.
> 

Got it, the channels feature is actually pretty cool.

> The other parts sound fine. I'll be interested to see the finished product!

Absolutely! Will be sharing the product and my experience in the `built-with-vendure` thread.

Thanks a ton for your help.

**Reply by rohanrajpal** [reply]
Hey

How do you suggest I store the tenant's login data? As the login would be through an OTP every time, I won't be storing any password. However, there will be a channel token associated with every token.

So I was thinking of keeping the login data directly in my Postgres and store the channel token along with the phone number in each row. I anyhow would be coding the logic of OTP login on my own, so I guess I don't need vendure to login/register the tenant? (Apart from changing/creating the channel)

Would appreciate any input.

**Reply by michaelbromley** [reply]
Hi!

It's hard for me to advise in detail without fully understanding the way your auth should work, but in general you should be aiming to work within the pattern described here: https://www.vendure.io/docs/developer-guide/authentication/

You could, for example, store the phone # as the `externalIdentifier` using the [`ExternalAuthenticationService`](https://www.vendure.io/docs/typescript-api/auth/external-authentication-service/), and then store the OTP and any other data you need to store in the `user.authenticationMethod.metadata` field.

**Reply by rohanrajpal** [reply]
Ohh, missed your reply somehow! Thanks for sharing this, will have a look and revert.

**Comment by rohanrajpal** [comment]
Good Morning Michael

I was trying to figure out the right way to use a channel with Nextjs Dyanmic routing.

I want to achieve the following with graphql, nextjs and urql SSR

1. I have setup a dynamic route [store-name].tsx
2. Now the store-name will be the value of the `vendure-token` token in my graphql url
3. This token will now be used for any of the future graphql requests

SSR is setup by following the guide https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#nextjs

```ts
/ pages/index.js
import React from 'react';
import Head from 'next/head';
import { withUrqlClient } from 'next-urql';
const Index = () => {
  const [result] = useQuery({
    query: '{ test }',
  });
  // ...
};

// was wondering if I could add the store-name to the url below
export default withUrqlClient((_ssrExchange, ctx) => ({
  // ...add your Client options here
  url: 'http://localhost:3000/graphql',
}))(Index);
```

I asked this question in the Urql, their solution works but they also suggested to use a better method to do the same
https://github.com/FormidableLabs/urql/discussions/1237

So now i'm thinking to use the channel token as a query parameter, but I cant figure out where exactly shall the query parameter go? The options in any of the queries does not seem to have a channel token parameter.

**Reply by michaelbromley** [reply]
Hi Rohan,

I've never used urql, so I can't advise on the specifics of how to implement this, but in the answer on the url discussion, they talk about adding it as a query param rather than a header, which is strange, since you clearly ask how to add a header. I see you later re-iterate you want to set a header. Let's see if they can advise on that.

**Reply by rohanrajpal** [reply]
Hey

Could you please suggest any alternative to urql? I'm making something for a hackathon and just want to make something functional for now. My only requirement is that it should work with `graphql-codegen.` There are quite a few options and I'm a little confused

Also does something like this exist? 
```graphql
query {
  products (vendure-token: "store-name") {
    items{
      id
      name
    }
  }
}
```

**Reply by michaelbromley** [reply]
I use Apollo Client. Works great with codegen. In fact, they even have a codegen plugin which will generate React code for you: https://graphql-code-generator.com/docs/plugins/typescript-react-apollo

> Also does something like this exist?

No, the channel token must be provided as an HTTP header or a query string only.

**Reply by rohanrajpal** [reply]
Amazing, thank you!

**Reply by rohanrajpal** [reply]
Hey @michaelbromley, peeps at urql helped me out to set the token header
https://github.com/FormidableLabs/urql/discussions/1237#discussioncomment-220460

Now for the Dashboard, I'm using client side rendering and am currently trying to implement the following flow
1. User signs up with Firebase Auth
2. Vendure token for the user is stored in Firestore
3. Now whenever this user signs in, I'll retrieve their token from Firestore and set it in my Urql client
4. Any subsequent requests for that user would now be only for their channel

I'm confused about the third part, could you let me know how did you do the same in the Admin UI and Apollo? I'll figure out to convert the Apollo logic to Urql, they're pretty similar.

**Reply by rohanrajpal** [reply]
I did some research and I guess I should use cookies to do the above?

**Reply by michaelbromley** [reply]
Cookies are very simple because the browser handles it all for you. Bearer tokens are useful when you e.g. need to support mobile apps or other non-browser clients. So yeah, use cookies if you only need to support the browser.

**Reply by rohanrajpal** [reply]
Got it. Thanks Michael and Merry Christmas :christmas_tree:

**Comment by elgadir** [comment]
Hello ;
And in the case we management the roles modules ( add products related with manager role who can controle manay channels)
Using  this methode
https://tech.canyonlegal.com/multitenancy-with-nestjs-typeorm-postgres

**Comment by kinthaiofficial** [comment]
Multi-tenancy in Vendure is achievable but requires some thought about where to draw the isolation boundary. A few approaches that work:

**Approach 1: Channel-based (built-in)**
Vendure's Channel system is designed for this. Each tenant gets a Channel, and products/orders/customers are scoped per-channel:

```typescript
// Create a channel per tenant
const channel = await channelService.create(ctx, {
  code: `tenant-${tenantId}`,
  token: generateToken(),
  currencyCode: CurrencyCode.USD,
  defaultLanguageCode: LanguageCode.en,
  pricesIncludeTax: false,
});

// All queries automatically scoped when you use the channel token
// API: Authorization: Bearer <channel-token>
```

Limitation: channels share the same database, so a tenant with millions of products will affect others.

**Approach 2: Schema-based isolation (Postgres)**
More isolation but more operational complexity:

```typescript
// Custom plugin: set search_path per request
@Injectable()
export class TenantPlugin implements NestInterceptor {
  intercept(context: ExecutionContext, next: CallHandler) {
    const tenantId = extractTenantId(context);
    return this.dataSource.query(`SET search_path TO tenant_${tenantId}`)
      .then(() => next.handle());
  }
}
```

**Approach 3: Separate Vendure instances**
Each tenant gets their own Vendure instance. Most isolation, most ops overhead. Worth it if tenants have substantially different configs or compliance requirements.

For most cases, Channel-based works well and avoids operational complexity. The main limitation is shared compute — if one tenant's storefront gets a traffic spike, it can affect others.

What's the scale and isolation requirement you're targeting?
